src/app/CommandHandler.h

```diff
@@ -319,6 +299,48 @@ class CommandHandler
      */
     FabricIndex GetAccessingFabricIndex() const;
 
+    /**
+     * @brief Best effort to add InvokeResponse to InvokeResponseMessage.
+     * 
+     * Tries to add response using lambda. Upon failure to add response, attempts
+     * to rollback the InvokeResponseMessage to a known good state. If failure is due
+     * to insufficient space in the current InvokeResponseMessage:
+     *  - Finalizes the current InvokeResponseMessage.
+     *  - Allocates a new InvokeResponseMessage.
+     *  - Reattempts to add the InvokeResponse to the new InvokeResponseMessage.
+     *    the new InvokeResponseMessage.
+     * 
+     * @param addResponseFunction: A lambda function responsible for adding the
+     *      response to the current InvokeResponseMessage.
+     */
+    template <typename Function>
+    CHIP_ERROR TryAddingResponse(Function && addResponseFunction)
+    {
+        mRollbackBackupValid = false;
+        CHIP_ERROR err       = addResponseFunction();
+        if (err == CHIP_NO_ERROR)
+        {
+            return CHIP_NO_ERROR;
+        }
+        ReturnErrorOnFailure(RollbackResponse());
+        // If we failed to an additional command due to lack of space in the
+        // packet, we will make another attempt to add the response using
+        // an additional InvokeResponseMessage.
+        if (mState != State::AddedCommand || err != CHIP_ERROR_NO_MEMORY)
+        {
+            return err;
+        }
+        ReturnErrorOnFailure(FinalizeInvokeResponseMessageAndReadyNext());
+        err                  = addResponseFunction();
+        mRollbackBackupValid = false;
+        if (err != CHIP_NO_ERROR)
+        {
+            // Over here we don't care about the return
+            RollbackResponse();
```

Comment:
This doesn't make sense to me.  We set mRollbackBackupValid to false above, so this will be a no-op, no?

I think these manual mRollbackBackupValid manipulations are pretty odd and at the very least need to be clearly documented.

Reply:
It should be. I will try adding unit test that make sure it would have caught this issue.
```
        mRollbackBackupValid = false;
        err                  = addResponseFunction();
```

Reply:
Unfortunately an attempt to encapsulate this and move it to the stack didn't work out :/
Working on adding documentation now